File: fix_encoding.js
const fs = require('fs');
const path = './client/src/components/Form20Tables.tsx';

// Read the file as binary (buffer)
const buf = fs.readFileSync(path);

// The file currently contains the string "àºžàº²àºªàº²" encoded as utf-8.
// We need to read it as utf-8 string, then take the char codes (which represent the corrupted latin1 bytes)
// and put them into a new buffer, then decode that new buffer as utf-8.
const str = buf.toString('utf8');

const correctedBytes = new Uint8Array(str.length);
let isCorrupted = false;
let j = 0;

for (let i = 0; i < str.length; i++) {
  const code = str.charCodeAt(i);
  if (code > 255) {
    // If there are actual Lao characters, it means it's not purely mojibake
    // Or it means the file is mixed.
  }
  // Convert char code back to byte
  correctedBytes[i] = code & 0xFF;
}

try {
  // If the whole file was double-encoded:
  // Let's try converting the corrupted string to a buffer using latin1, then reading as utf8
  const latin1Buf = Buffer.from(str, 'latin1');
  const fixedStr = latin1Buf.toString('utf8');
  
  if (fixedStr.includes('ພາສາ')) {
     fs.writeFileSync(path, fixedStr, 'utf8');
     console.log('Successfully fixed encoding!');
  } else {
     console.log('Failed to fix: did not find expected Lao words after decode.');
  }
} catch (e) {
  console.log('Error:', e.message);
}

File: fix_encoding.py
import os

file_path = 'client/src/components/Form20Tables.tsx'

with open(file_path, 'r', encoding='utf-8') as f:
    content = f.read()

# Encode the string to bytes using windows-1252 (which maps the mojibake chars to original bytes)
try:
    original_bytes = content.encode('cp1252')
    # Decode the bytes using utf-8 to get the proper Lao characters
    fixed_content = original_bytes.decode('utf-8')
    
    if 'ພາສາ' in fixed_content:
        with open(file_path, 'w', encoding='utf-8') as f:
            f.write(fixed_content)
        print('Successfully fixed encoding using cp1252!')
    else:
        print('Failed to fix: "ພາສາ" not found in decoded content.')
except Exception as e:
    print('Error:', e)
